Add CancellationToken Support to Mediator Pipeline and Handlers (#5)

luylucas10 opened the pull request:
This PR implements full CancellationToken support throughout the OpenMediator pipeline as discussed in [issue #5]

- Added: CancellationToken as an optional parameter to all relevant async methods on IMediatorBus, middleware, and handler interfaces.
- Propagated: The cancellation token throughout the middleware pipeline and to all registered handlers.
- Refactored: Simplified interface signatures using default values instead of overloads.
- Updated: Unit tests to reflect the new signatures, and added new tests to ensure proper cancellation propagation and OperationCanceledException throwing.
- Ensured: Backward compatibility for existing consumers using the default CancellationToken parameter.
- Fixed typo: Corrected TRespose to TResponse in the generic handler interface.

Notes

- All tests are passing, including new scenarios validating cancellation logic.
- This change improves compliance with .NET best practices and prepares the library for cloud-native and long-running operations.
Closes [#5]

Let me know if you need any further changes or documentation improvements! 😊

luysantanadev committed:
feat: add CancellationToken support to mediator pipeline and handlers (#5)

- Add CancellationToken as optional parameter to IMediatorBus, middleware, and handler interfaces
- Propagate token through pipeline and handlers
- Refactor method signatures to use default parameter values instead of overloads
- Update and add tests to verify cancellation is respected (OperationCanceledException)
- Fix generic type typo (TRespose -> TResponse)
- All tests passing

Closes #5

Sergi0Martin commented on src/Buses/DefaultMediatorBus.cs:
```diff
@@ -1,31 +1,38 @@
+﻿using System.Runtime.CompilerServices;
+using OpenMediator.Middlewares;
 
+[assembly: InternalsVisibleTo("OpenMediator.Unit.Test")]
```
Add InternalsVisibleTo at project level (OpenMediator.csproj) Instead of in a class. This way:

``` csharp
<ItemGroup>
	<AssemblyAttribute Include="System.Runtime.CompilerServices.InternalsVisibleTo">
		<_Parameter1>a3innuva.Accounting.Api.Query.Host.Integration.Tests</_Parameter1>
	</AssemblyAttribute>
</ItemGroup>
```

Sergi0Martin commented on src/Buses/DefaultMediatorBus.cs:
```diff
@@ -1,31 +1,38 @@
     IEnumerable<IMediatorMiddleware> _middlewares) : IMediatorBus
 {
-    public async Task<TResponse> SendAsync<TCommand, TResponse>(TCommand request)
+    public async Task<TResponse> SendAsync<TCommand, TResponse>(TCommand request,
```
One param by line
``` csharp
public async Task<TResponse> SendAsync<TCommand, TResponse>(
    TCommand request,
    CancellationToken cancellationToken = default)
```

luylucas10 commented:
Nice, I'll apply this soon.

Sergi0Martin commented on src/Buses/DefaultMediatorBus.cs:
```diff
@@ -1,31 +1,38 @@
-        var middlewareTask = _middlewares.Aggregate(() => next(), (nextMiddleware, middleware) => async () => await middleware.ExecuteAsync(command, nextMiddleware));
+        var middlewareTask = _middlewares.Aggregate(() => next(),
+            (nextMiddleware, middleware) => async () =>
+                await middleware.ExecuteAsync(command, nextMiddleware, cancellationToken));
```
change to inline for readability

Sergi0Martin commented on test/OpenMediator.Unit.Test/Commands/CommandTest.cs:
```diff
@@ -44,4 +44,21 @@ public async Task Send_Command_Throw_If_Not_Registered()
+        var command = new LongRunningCommand();
+
+        using var cts = new CancellationTokenSource();
+        await cts.CancelAsync(); // Já está cancelado antes de executar
```
```suggestion
        await cts.CancelAsync(); // Cancelled before execution
```

Sergi0Martin reviewed (changes requested):
I let you some feedback. Thanks for your contribution. 😁

luysantanadev committed:
refactor: format method signatures for improved readability

Set one param per line in classes:
- DefaultMediatorBus.cs
- IMediatorBus.cs
- IMediatorMiddleware.cs
- ICommandHandler.cs

luysantanadev committed:
refactor: streamline middleware task aggregation for improved clarity

luylucas10 commented:
Thanks, I made all the refactors that you suggest. Let me know if there are more things to improve.

Thank you to create this library, It's simple and really usefull.
I like the mediator pattern 'cause I don't need to worry about injections anymore.

Sergi0Martin committed:
Upgrade version

Sergi0Martin commented:
> Thanks, I made all the refactors that you suggest. Let me know if there are more things to improve.
> 
> Thank you to create this library, It's simple and really usefull. I like the mediator pattern 'cause I don't need to worry about injections anymore.

I just pushed one commit because I forgot to mention the need of upgrade version. Thanks for your contribution @luylucas10

Sergi0Martin reviewed (approved):
good job 🚀
